client/tracing/src/block/mod.rs

```diff
@@ -0,0 +1,314 @@
+use sc_rpc_server::MAX_PAYLOAD;
+use sp_api::{Core, Metadata, ProvideRuntimeApi, Encode};
+use sp_blockchain::HeaderBackend;
+use sp_runtime::{
+	generic::BlockId,
+	traits::{Block as BlockT, Header},
+};
+use sp_rpc::tracing::{BlockTrace, Span, TraceError, TraceBlockResponse};
+use sp_tracing::{WASM_NAME_KEY, WASM_TARGET_KEY, WASM_TRACE_IDENTIFIER};
+use sp_core::hexdisplay::HexDisplay;
+use crate::{SpanDatum, TraceEvent, Values};
+
+// Default to only pallet, frame support and state related traces
+const DEFAULT_TARGETS: &'static str = "pallet,frame,state";
+const TRACE_TARGET: &'static str = "block_trace";
+// Default to only events with the key prefixes for :extrinsic_index & system::Account.
+const DEFAULT_STORAGE_KEYS: &'static str =
+	"3a65787472696e7369635f696e646578,26aa394eea5630e07c48ae0c9558cef7b99d880ec681799c0cf30e8886371da9";
+
+struct BlockSubscriber {
+	targets: Vec<(String, Level)>,
+	next_id: AtomicU64,
+	current_span: CurrentSpan,
+	spans: Mutex<HashMap<Id, SpanDatum>>,
+	events: Mutex<Vec<TraceEvent>>,
+}
+
+impl BlockSubscriber {
+	fn new(
+		targets: &str,
+		spans: Mutex<HashMap<Id, SpanDatum>>,
+		events: Mutex<Vec<TraceEvent>>,
+	) -> Self {
+		let next_id = AtomicU64::new(1);
+		let mut targets: Vec<_> = targets
+			.split(',')
+			.map(|s| crate::parse_target(s))
+			.collect();
+		// Ensure that WASM traces are always enabled
+		// Filtering happens when decoding the actual target / level
+		targets.push((WASM_TRACE_IDENTIFIER.to_owned(), Level::TRACE));
+		BlockSubscriber {
+			targets,
+			next_id,
+			current_span: CurrentSpan::default(),
+			spans,
+			events,
+		}
+	}
+}
+
+impl Subscriber for BlockSubscriber {
+	fn enabled(&self, metadata: &tracing::Metadata<'_>) -> bool {
+		for (target, level) in &self.targets {
+			if metadata.target().starts_with(target.as_str()) && metadata.level() <= level {
+				return true;
+			}
+		}
+		false
+	}
+
+	fn new_span(&self, attrs: &Attributes<'_>) -> Id {
+		let id = self.next_id.fetch_add(1, Ordering::Relaxed);
+		let id = Id::from_u64(id);
+		let mut values = Values::default();
+		attrs.record(&mut values);
+		let parent_id = attrs.parent().cloned()
+			.or_else(|| self.current_span.id());
+		let span = SpanDatum {
+			id: id.clone(),
+			parent_id,
+			name: attrs.metadata().name().to_owned(),
+			target: attrs.metadata().target().to_owned(),
+			level: *attrs.metadata().level(),
+			line: attrs.metadata().line().unwrap_or(0),
+			start_time: Instant::now(),
+			values,
+			overall_time: Default::default()
+		};
+
+		self.spans.lock().insert(id.clone(), span);
+		id
+	}
+
+	fn record(&self, span: &Id, values: &Record<'_>) {
+		let mut span_data = self.spans.lock();
+		if let Some(s) = span_data.get_mut(span) {
+			values.record(&mut s.values);
+		}
+	}
+
+	fn record_follows_from(&self, _span: &Id, _follows: &Id) {
+		// Not currently used
+	}
+
+	fn event(&self, event: &tracing::Event<'_>) {
+		let mut values = crate::Values::default();
+		event.record(&mut values);
```

Comment:
Same q as above with `attrs.record` in `new_span`

Reply:
Yes, this one we need because it's where all your storage keys etc are recorded